vm/starnix/starnix.go

```diff
@@ -0,0 +1,383 @@
+	if inst.debug {
+		log.Logf(0, "running command: %s %q", cmd, args)
+	}
+	output, err := osutil.RunCmd(timeout, inst.fuchsiaDirectory, cmd, args...)
+	if inst.debug {
+		log.Logf(0, "%s", output)
+	}
+	return err
+}
+
+func (inst *instance) Forward(port int) (string, error) {
+	if port == 0 {
+		return "", fmt.Errorf("vm/starnix: forward port is zero")
+	}
+	return fmt.Sprintf("localhost:%v", port), nil
+}
+
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	base := filepath.Base(hostSrc)
+	vmDst := filepath.Join(inst.targetDir, base)
+	if base == "syz-fuzzer" || base == "syz-execprog" {
+		return hostSrc, nil // we will run these on host.
+	}
+
+	err := inst.runCommand(
+		1*time.Minute*inst.timeouts.Scale,
+		"adb",
+		"-s",
+		fmt.Sprintf("127.0.0.1:%d", inst.port),
+		"push",
+		hostSrc,
+		vmDst)
+	return vmDst, err
+}
+
+func (inst *instance) Run(timeout time.Duration, stop <-chan bool, command string) (
+	<-chan []byte, <-chan error, error) {
+	rpipe, wpipe, err := osutil.LongPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	inst.merger.Add("adb", rpipe)
+
+	args := strings.Split(command, " ")
+	if bin := filepath.Base(args[0]); bin == "syz-fuzzer" || bin == "syz-execprog" {
+		for i, arg := range args {
+			if strings.HasPrefix(arg, "-executor=") {
+				args[i] = fmt.Sprintf("-executor=%s %d", inst.executor, inst.port)
+				// TODO(fxbug.dev/120202): reenable threaded mode once clone3 is fixed.
+				args = append(args[:i+2], args[i+1:]...)
+				args[i+1] = "-threaded=0"
+			}
+		}
+	}
+	if inst.debug {
+		log.Logf(0, "running command: %#v", args)
+	}
+	cmd := osutil.Command(args[0], args[1:]...)
+	cmd.Dir = inst.workdir
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	if err := cmd.Run(); err != nil {
+		wpipe.Close()
+		return nil, nil, err
+	}
+	wpipe.Close()
+	errc := make(chan error, 1)
+	signal := func(err error) {
+		select {
+		case errc <- err:
+		default:
+		}
+	}
+
+	go func() {
+	retry:
+		select {
+		case <-time.After(timeout):
+			signal(vmimpl.ErrTimeout)
+		case <-stop:
+			signal(vmimpl.ErrTimeout)
+		case <-inst.diagnose:
+			cmd.Process.Kill()
+			goto retry
+		case err := <-inst.merger.Err:
+			cmd.Process.Kill()
+			if cmdErr := cmd.Wait(); cmdErr == nil {
+				// If the command exited successfully, we got EOF error from merger.
+				// But in this case no error has happened and the EOF is expected.
+				err = nil
+			}
+			signal(err)
+			return
+		}
+		cmd.Process.Kill()
+		cmd.Wait()
+	}()
+	return inst.merger.Output, errc, nil
+}
+
+func (inst *instance) Info() ([]byte, error) {
+	info := fmt.Sprintf("%v\n%v %q", inst.version, "ffx", inst.args)
+	return []byte(info), nil
+}
+
+func (inst *instance) Diagnose(rep *report.Report) ([]byte, bool) {
+	return nil, false
+}
+
+func (inst *instance) setFuchsiaVersion() error {
+	version, err := osutil.RunCmd(1*time.Minute*inst.timeouts.Scale, inst.fuchsiaDirectory, inst.ffxBinary, "version")
```

Comment:
This better belongs to the pool ctor. We don't need to run it for each instance/boot. And if ffx fails on such simple commands, we better report early and exit.